---
title: "RepeatModeler Timema"
output:
  html_document: default
  pdf_document: default
  word_document: default
---

```{r setup, include=FALSE}
library(dplyr)
library(ggplot2)

knitr::opts_chunk$set(echo = TRUE)
```

# RepeatModeler

RepeatModeler is run on each assembly to get a fasta of consensus sequences for each genome.

```{bash, eval=F}

mkdir /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM


cd /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM

for Path in `ls /work/FAC/FBM/DEE/tschwand/default/vmerel/GPT/Assemblies/*fasta`
do

  File=`echo $Path | cut -f 11 -d '/'`
  ID=`echo $File | sed 's/\(.*\)_.*/\1/'`
  echo $File
  echo $ID
  rm -r /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/$ID
  mkdir /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/$ID
  cd /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/$ID
  cp $Path ./

  echo '#!/bin/bash -l

#SBATCH --account=tschwand_timema_lr_genomes
#SBATCH --mail-type NONE 
#SBATCH --mail-user vincent.merel@unil.ch

#SBATCH --chdir /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/'"$ID"'
#SBATCH --job-name RepeatModeler_'"$ID"'
#SBATCH --output RepeatModeler_'"$ID"'.out
#SBATCH --error RepeatModeler_'"$ID"'.err

#SBATCH --partition cpu

#SBATCH --nodes 1 
#SBATCH --ntasks 1 
#SBATCH --cpus-per-task 40
#SBATCH --mem 120G 
#SBATCH --time 72:00:00 
#SBATCH --export NONE

module load singularity
singularity exec --bind /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/'"$ID"':/work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/'"$ID"' \
--pwd /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/'"$ID"' \
docker://dfam/tetools:latest \
BuildDatabase -name genome '"$File"'

singularity exec --bind /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/'"$ID"':/work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/'"$ID"' \
--pwd /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/'"$ID"' \
docker://dfam/tetools:latest \
RepeatModeler -database genome \
-pa 40 \
-LTRStruct >& run.out' > /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/$ID/RepeatModeler_$ID.sh

sbatch /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/$ID/RepeatModeler_$ID.sh

done


```

# Cluster

To reduce redundancy all the generated consensus are clustered using a 80% of identity threshold.

```{bash, eval=F}

cd /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs

rm seqDb.fas
#Let's not consider Tce_NA
Samples="Tbi Tce Tce_N Tcm Tdi Tge_A Tge_H20 Tms Tpa Tps Tsi"

#Add a different header to consensus for each species
for Sample in $Samples
do
  echo "###"
  echo $Sample
  echo "###"
  sed 's/\#/_'"$Sample"'\#/g' /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/RM/$Sample/genome-families.fa >>  seqDb.fas
done

sed 's/\#/_Tte\#/g' /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/HiFi_assembly/Tte/results/Tte/flye/repeatmodeler/genome-families.fa >> seqDb.fas


#A big fa
mkdir /scratch/vmerel/mmseqs
 echo '#!/bin/bash -l

#SBATCH --account=tschwand_timema_lr_genomes
#SBATCH --mail-type NONE 
#SBATCH --mail-user vincent.merel@unil.ch

#SBATCH --chdir /scratch/vmerel/mmseqs
#SBATCH --job-name mmseqs
#SBATCH --output /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/mmseqs.out
#SBATCH --error /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/mmseqs.err

#SBATCH --partition cpu

#SBATCH --nodes 1 
#SBATCH --ntasks 1 
#SBATCH --cpus-per-task 20
#SBATCH --mem 60G 
#SBATCH --time 00:10:00 
#SBATCH --export NONE

module load gcc/10.4.0 mmseqs2/13-45111

mmseqs createdb /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/seqDb.fas seqDb

mmseqs cluster seqDb clusterDb tmp \
--min-seq-id 0.8 \
--threads 20 \
-k 0 \
--cov-mode 1

mmseqs result2repseq seqDb clusterDb clu_rep
mmseqs result2flat seqDb seqDb clu_rep RM.fasta --use-fasta-header
cp RM.fasta /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/' > /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/mmseqs.sh
sbatch /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/mmseqs.sh
```

# Classification

This command line is just converting RepeatModeler fasta header to a .csv classification with sequence name, family and superfamily 

```{bash, eval=F}

grep '>' /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/RM.fasta |\
sed 's/>//g' |\
awk -F'[/# (]' '{
if ($2=="Unknown")
  print $1 "\tUnknown\tUnknown";
else if ($2=="tRNA" || $2=="Simple_repeat" || $3=="")
        print $1 "\t" $2 "\t" $2;
    else 
        print $1 "\t" $2 "\t" $3;
}' > /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/RM.classi
```



# RepeatMasker

Running RepeatMasker to annotate Tdi genome

```{bash, eval=F}

rm -r /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/RepeatMasker
mkdir /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/RepeatMasker
cd /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/RepeatMasker
#cp  /nas/FAC/FBM/DEE/tschwand/timema_lr_genomes/D2c/timema_genomes/fasta/Tte_LRv1.fasta /work/FAC/FBM/DEE/tschwand/default/vmerel/GPT/Assemblies/

cd ~

for Path in `ls /work/FAC/FBM/DEE/tschwand/default/vmerel/GPT/Assemblies/Tdi*fasta`
do

  File=`echo $Path | cut -f 11 -d '/'`
  ID=`echo $File | sed 's/\(.*\)_.*/\1/'`
  echo $File
  echo $ID

  rm -r /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/RepeatMasker/$ID
  mkdir /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/RepeatMasker/$ID
  
  echo '#!/bin/bash -l

#SBATCH --account=tschwand_default
#SBATCH --mail-type NONE 
#SBATCH --mail-user vincent.merel@unil.ch

#SBATCH --chdir /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/RepeatMasker/'"$ID"'
#SBATCH --job-name RepeatMasker_'"$ID"'
#SBATCH --output RepeatMasker_'"$ID"'.out
#SBATCH --error RepeatMasker_'"$ID"'.err

#SBATCH --partition cpu

#SBATCH --nodes 1 
#SBATCH --ntasks 1 
#SBATCH --cpus-per-task 32
#SBATCH --mem 64G 
#SBATCH --time 70:00:00 
#SBATCH --export NONE

date

module load gcc/10.4.0 repeatmasker/4.1.2-p1

#-s \ -cutoff 200 \  -gccalc \ 

RepeatMasker -gff \
-u \
-pa 8 \
-no_is \
-lib /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/RM.fasta \
'"$Path"' 

date' > /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/RepeatMasker/$ID/RepeatMasker_$ID.sh

sbatch /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/RepeatMasker/$ID/RepeatMasker_$ID.sh

done
```

# .bed for Tdi

Creating a bedlike file from RepeatMasker output and incorporating the classification 

```{bash, eval=F}

#A bed file without classification as a start
#Will add the classification after

mkdir /tmp/Will
cp /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/RM.fasta /tmp/Will/.
grep -v "##" /work/FAC/FBM/DEE/tschwand/default/vmerel/GPT/Assemblies/Tdi_LRv5a.fasta.out.gff | awk 'OFS="\t" {print $1,$4-1,$5, $10}' |  sed -E 's/Motif:|"//g' | sort -k1,1 -k2,2n > /tmp/Will/Tdi_AllRepeats.bed
#| bedtools merge 
wc -l /tmp/Will/Tdi_AllRepeats.bed
wc -l /work/FAC/FBM/DEE/tschwand/default/vmerel/GPT/Assemblies/Tdi_LRv5a.fasta.out.gff

sort -k4,4 /tmp/Will/Tdi_AllRepeats.bed > /tmp/Will/Tdi_AllRepeats.sorted.bed
sort -k1,1 /work/FAC/FBM/DEE/tschwand/timema_lr_genomes/vmerel/GPT/Repeats/mmseqs/RM.classi > /tmp/Will/RM.sorted.classi

#Adding the classification
join -1 4 \
-2 1 \
/tmp/Will/Tdi_AllRepeats.sorted.bed \
/tmp/Will/RM.sorted.classi |  awk '{print $2,$3,$4,$1,$5,$6}' | tr ' ' '\t' | sort -k1,1 -k2,2n >  /tmp/Will/Tdi_AllRepeats.classi.bed
```
